File: sprint02/JMS/PGS_43164.cpp
#include <string>
#include <vector>
#include <cstring>
#include <algorithm>
#include <iostream>

using namespace std;

vector<string> answer;
int visited[10001] = {0};

bool DFS(string dest, vector<vector<string>> tickets){
    answer.push_back(dest);
    
    for(int i=0;i<tickets.size();i++){
        if(!visited[i] && !tickets[i][0].compare(dest)){
            visited[i]  = 1;
            if(DFS(tickets[i][1], tickets))
                return true;
            else
                visited[i] = 0;
        }
    }
    
    if(answer.size() - 1 == tickets.size())
        return true;
    else{
        answer.pop_back();
        return false;
    }
}

vector<string> solution(vector<vector<string>> tickets) {
    sort(tickets.begin(), tickets.end());
    
    DFS("ICN", tickets);
    
    return answer;
}

File: sprint02/KKW/PGS_43164.py
def solution(tickets):
    answer = []
    airline = {}
    for t in tickets:
        if t[0] in airline:
            airline[t[0]].append(t[1])
        else:
            airline[t[0]] = [t[1]]

    for a in airline.keys():
        airline[a].sort(reverse=True)

    start = ["ICN"]
    while start:
        target = start[-1]
        if target not in airline or len(airline[target]) == 0:
            answer.append(start.pop())
        else:
            start.append(airline[target].pop())

    return answer[::-1]